src/main/scala/is/hail/asm4s/FunctionBuilder.scala

```diff
@@ -55,8 +55,66 @@ object FunctionBuilder {
     new FunctionBuilder(Array(GenericTypeInfo[A], GenericTypeInfo[B], GenericTypeInfo[C], GenericTypeInfo[D], GenericTypeInfo[E]), GenericTypeInfo[R])
 
   private implicit def methodNodeToGrowable(mn: MethodNode): Growable[AbstractInsnNode] = new Growable[AbstractInsnNode] {
-    def +=(e: AbstractInsnNode) = { mn.instructions.add(e); this }
-    def clear() { throw new UnsupportedOperationException() }
+    def +=(e: AbstractInsnNode) = {
+      mn.instructions.add(e);
+      this
+    }
+
+    def clear() {
+      throw new UnsupportedOperationException()
+    }
+  }
+}
+
+class MethodBuilder(val fb: FunctionBuilder[_], mname: String, parameterTypeInfo: Array[MaybeGenericTypeInfo[_]], returnTypeInfo: MaybeGenericTypeInfo[_]) {
+
+  def descriptor: String = s"(${ parameterTypeInfo.map(_.base.name).mkString })${ returnTypeInfo.base.name }"
+
+  val mn = new MethodNode(ACC_PUBLIC, mname, descriptor, null, null)
+  fb.cn.methods.asInstanceOf[util.List[MethodNode]].add(mn)
+
+  val start = new LabelNode
+  val end = new LabelNode
+  val layout: Array[Int] = 0 +: (parameterTypeInfo.scanLeft(1) { case (prev, gti) => prev + gti.base.slots })
+  val argIndex: Array[Int] = layout.init
+  var locals: Int = layout.last
+
+  def allocLocal[T](name: String = null)(implicit tti: TypeInfo[T]): Int = {
+    val i = locals
+    assert(i < 65536)
+    locals += tti.slots
+
+    mn.localVariables.asInstanceOf[util.List[LocalVariableNode]]
+      .add(new LocalVariableNode(if (name == null) "local" + i else name, tti.name, null, start, end, i))
+    i
+  }
+
+  def newLocal[T](implicit tti: TypeInfo[T]): LocalRef[T] =
+    newLocal()
+
+  def newLocal[T](name: String = null)(implicit tti: TypeInfo[T]): LocalRef[T] =
+    new LocalRef[T](allocLocal[T](name))
+
+  def getArg[T](i: Int)(implicit tti: TypeInfo[T]): LocalRef[T] = {
+    assert(i >= 0)
+    assert(i < layout.length)
+    new LocalRef[T](argIndex(i))
+  }
+
+  val l = new mutable.ArrayBuffer[AbstractInsnNode]()
+
+  def emit(c: Code[_]) {
+    c.emit(l)
+  }
+
+  def emit(insn: AbstractInsnNode) {
+    l += insn
+  }
+
+  def invoke[R: TypeInfo](args: Code[_]*)(implicit rti :TypeInfo[R]): Code[R] = {
```

Comment:
one way to make this slightly nicer is to subclass it for arities 0, 1, and 2 (the common ones) and provide an `invoke` with the right number of arguments (and then you can get the types right too by taking the right number of typeinfos) that just calls this one.